src/Lucene.Net/Index/ArrayHolder.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Buffers;
+using System.Threading;
+using Lucene.Net.Store;
+
+namespace Lucene.Net.Index
+{
+    public class ArrayHolder : IDisposable
+    {
+        private readonly int _size;
+        private readonly Directory _directory;
+        private readonly string _name;
+        private readonly long[] _longArray;
+        private readonly Term[] _termArray;
+        private readonly TermInfo[] _termInfoArray;
+        private int _usages;
+
+        public Span<long> LongArray => _longArray.AsSpan(0, _size);
+        public Span<TermInfo> InfoArray => _termInfoArray.AsSpan(0, _size);
+        public Span<Term> IndexTerms => _termArray.AsSpan(0, _size);
+
+        public ArrayHolder(int size, Directory directory, string name)
+        {
+            _size = size;
+            _directory = directory;
+            _name = name;
+            _longArray = ArrayPool<long>.Shared.Rent(size);
+            _termArray = ArrayPool<Term>.Shared.Rent(size);
+            _termInfoArray = ArrayPool<TermInfo>.Shared.Rent(size);
+        }
+
+        public void AddRef()
+        {
+            Interlocked.Increment(ref _usages);
+        }
+
+        public void ReleaseRef()
+        {
+            if (Interlocked.Decrement(ref _usages) == 0)
+                _directory.RemoveFromTermsIndexCache(_name);
+        }
+
+        public static ArrayHolder GenerateArrayHolder(Directory directory, string name, FieldInfos fieldInfos, int readBufferSize, int indexDivisor, IState state)
+        {
+            var indexEnum = new SegmentTermEnum(directory.OpenInput(name, readBufferSize, state), fieldInfos, true, state);
+
+            try
+            {
+                int indexSize = 1 + ((int)indexEnum.size - 1) / indexDivisor; // otherwise read index
+
+                var holder = new ArrayHolder(indexSize, directory, name);
+
+                for (int i = 0; indexEnum.Next(state); i++)
+                {
+                    holder.IndexTerms[i] = indexEnum.Term;
+                    holder.InfoArray[i] = indexEnum.TermInfo();
+                    holder.LongArray[i] = indexEnum.indexPointer;
+
+                    for (int j = 1; j < indexDivisor; j++)
+                        if (!indexEnum.Next(state))
+                            break;
+                }
+
+                return holder;
+
+            }
+            finally
+            {
+                indexEnum?.Close();
+            }
+        }
+
+        public void Dispose()
+        {
+            GC.SuppressFinalize(this);
+
+            if (_size > 64 * 1024)
```

Comment:
wouldn't this case a lot of GC? IIRC shared pools take up to 1mb, rest is GC'ed

Reply:
I guess that this is here:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L25

?

Reply:
https://adamsitnik.com/Array-Pool/#how-to-use-it I remember it from here, but it looks it is 1gb now?

Reply:
increased to `256K`